library/std/src/panicking.rs

```diff
@@ -571,7 +581,16 @@ fn rust_panic_with_hook(
     // process real quickly as we don't want to try calling it again as it'll
     // probably just panic again.
     if panics > 2 {
-        util::dumb_print(format_args!("thread panicked while processing panic. aborting.\n"));
+        if (panics & !panic_count::ALWAYS_ABORT_FLAG) > 2 {
+            // Don't try to print the message in this case
+            // - perhaps that is causing the recursive panics.
+            util::dumb_print(format_args!("thread panicked while processing panic. aborting.\n"));
+        } else {
+            if let Some(message) = message {
+                util::dumb_print(*message);
+            }
```

Comment:
It'd be good to display the message and location in the same way as the regular panics.

If you construct a `PanicInfo` with `PanicInfo::internal_constructor(message, location)`, you can format that `PanicInfo` instead of just the message. It has a `Display` implementation that shows `panicked at '{message}', {location}` or `panicked at {location}` depending on whether a message is available.

(Then you could also combine both dump_print calls here into one, since `PanicInfo` will take care of handling the `message = None` case.)

Reply:
> It'd be good to display the message and location in the same way as the regular panics.
> 
> If you construct a `PanicInfo` with `PanicInfo::internal_constructor(message, location)`
...
> (Then you could also combine both dump_print calls here into one, since `PanicInfo` will take care of handling the `message = None` case.)

Thanks for the suggestion, which I have incorporated.  Ideally we would print a backtrace but AFAICT we cannot do that without allocating, with the existing `Backtrace` api.